Feat/implement magic link authorization

meczajkowski opened the pull request:
Please test this in your environment and let me know if any changes need to be made. I can provide access to my remote database environment, allowing you to thoroughly test the changes. Your feedback is valuable in ensuring the robustness and reliability of the implemented features.

## Summary:

This PR introduces the capability for hunters to log in to the application. The task involved a significant refactoring of the authentication system due to the previously hardcoded GitHub authentication method. Now, the authentication system supports two providers, each tailored to a specific type of user.

## Changes Made:

- Refactored the authentication system to support multiple providers.
- Introduced two authentication providers: GitHub and email.
- Implemented role-based access control:
- Users logging in via GitHub are assigned the "SPECIALIST" role, enabling them to create a profile.
- Users logging in via email are designated as "HUNTERs."
- Implemented error handling:
- Specialized error handling prevents specialists from logging in with the email provider, displaying an error message and preventing the sending of a verification link.
- Redirected hunters attempting to log in via GitHub to a dedicated error page, indicating that such action is not permitted as per requirements.


## Additional Details:

- The refactored authentication system ensures clearer delineation between user roles and authentication providers, enhancing the overall user experience and system integrity.
- The introduced error handling mechanisms prevent unintended access attempts and ensure compliance with application requirements.
- Adding additional providers in the future will now be a straightforward process.

Sumick reviewed (changes requested):
- [x] GithubDetails is not being created when user signs up to GDH (screen). Such model such keep user Github related data
- [x] Currently it won't work as users don't have SPECIALIST role. We should also pass validation when users have USER role
- [x] I can create profile as HUNTER (screen) - probably we have to merge Szymon PR into that one

<img width="1345" alt="Zrzut ekranu 2024-03-25 o 19 23 45" src="https://github.com/nerdbord/good-dev-hunting-app/assets/22893188/9ed3b086-cdc4-4cc2-b5fa-cb60bc0e81f3">


<img width="1396" alt="Zrzut ekranu 2024-03-25 o 19 21 35" src="https://github.com/nerdbord/good-dev-hunting-app/assets/22893188/59f234e9-7520-4ae2-aa8d-a75abd77e590">

meczajkowski commented:
- Users with the HUNTER role are now restricted from creating profiles and accessing any profile routes.
- The signIn function has been updated to check if a user does not have the HUNTER role, ensuring backward compatibility for users with the SPECIALIST role.

Sumick reviewed (approved):
Code looks good, and it works fine as I tested it on clear DB - good job! We can merge it on staging if you add what you have left on your end. I will re-create staging with production database records tomorrow, and we will simulate how it will affect prod before we can release it.

meczajkowski commented on src/auth.ts:
```diff
@@ -0,0 +1,142 @@
+
+      let roleToAdd: Role = 'USER'
+      if (account?.provider === 'github') {
+        if (profile?.login && user.id) {
```
Error handling has been updated. Our custom messages are now displayed in the server console, ensuring better insight into potential issues.

mmcode9407 reviewed (commented):
I left some comments :)

meczajkowski commented:
- The new auth implementation now accommodates users previously created in production, allowing them to log in with GitHub just as they were doing before this PR.
- Session callback has been implemented to store all necessary user details. This enables us to transition away from querying the database in our server components and instead authenticate users directly from the session. This approach greatly enhances performance, ensures data consistency, and simplifies implementation, significantly reducing the number of database queries required.

Sumick commented on src/app/(auth)/(components)/HunterAcc/HunterAcc.tsx:
```diff
@@ -0,0 +1,28 @@
+import { useSession } from 'next-auth/react'
+import styles from './HunterAcc.module.scss'
+
+const HunterAcc = () => {
```
**nit:** Doesn't look like it needs to be client component

Sumick commented on src/app/(auth)/helpers.ts:
```diff
@@ -0,0 +1,41 @@
+import { Role, type User } from '@prisma/client'
+
+interface UserAuthed {
+  id: string
```
**nit:** These are unused types, we can remove it

meczajkowski replied:
done

Sumick commented on src/backend/user/user.service.ts:
```diff
@@ -83,23 +82,27 @@ export async function syncUserWithGithub(credentials: {
-  return createdUser
+  if (existingDetails) {
+    // GitHub details already exist for this user, so we don't need to create again
+    return
```
**nit:** Let's return found `existingDetails` instead of undefined

meczajkowski replied:
This method has been removed because the functionality is now handled by the custom createUser method.
Refer to the implementation in the following commit: https://github.com/nerdbord/good-dev-hunting-app/pull/263/commits/110e303c9f00429dda5cf014708327f86c3711be

Sumick commented on src/components/landing-page/LandingHeader/LandingHeader.module.scss:
```diff
@@ -88,21 +92,30 @@
+  @include mobile {
+    display: flex;
+    // align-items: center;
+    // gap: 8px;
```
**nit:** comment to remove

Sumick commented on src/components/landing-page/UseYourProfile/UseYourProfile.tsx:
```diff
@@ -1,18 +1,14 @@
-  const profile = session
-    ? await findProfileWithUserInclude(session.user.email)
-    : null
+  const { user } = await getAuthorizedUser()
```
**question:** What about moving user to session object? Do we want to split in different ticket?

meczajkowski replied:
This implementation has been added at https://github.com/nerdbord/good-dev-hunting-app/pull/263/commits/bc5586fbfed1a0473c20998a51af006fa9ab7330. It's working fine now. Our next step is to verify if there are any remaining queries to the database in other components.

meczajkowski replied:
done in https://github.com/nerdbord/good-dev-hunting-app/pull/263/commits/db88700f82eec9b896628c3f671298b5714baa82

Sumick reviewed (approved):
Last minor things to implement that I've caught

meczajkowski commented on src/app/(auth)/helpers.ts:
```diff
@@ -0,0 +1,41 @@
+  login: string
+  avatar_url: string
+}
+
```
done

meczajkowski commented on src/app/api/auth/[...nextauth]/route.ts:
```diff
@@ -1,5 +1,3 @@
-export { handler as GET, handler as POST }
+export { GET, POST } from '@/auth'
+// export { GET, POST } from '../../../../auth'
+// export const runtime = 'edge' // optional
```
done

meczajkowski commented:
Changes made since my last comment
- Adjusted the menu button to align with the new authentication system.
- Updated headers to accommodate changes in authentication mechanisms.
- Fixed the issue where non-users were not redirected to the login page when accessing profiles via Magic Link.
- Restored the UI banner for non-registered users.
- Added the HunterAcc component.
- Removed unused types in (auth)/helpers.ts.
- Removed comments from the LandingHeader and NextAuth route.
- Implemented the use of sessions to verify user status in the getAuthorizedUser function.

Sumick commented on src/app/(profile)/my-profile/(components)/CreateProfileForm/CreateProfileForm.tsx:
```diff
@@ -125,7 +125,9 @@ const CreateProfileForm = () => {
-        await createProfile(payload)
+        const createdProfile = await createProfile(payload)
+        createdProfile &&
+          updateSession({ ...session.user, name: payload.fullName })
```
**issue:** We are missing `profileId` here, and user will have profileId as null even if he has profile created

meczajkowski replied:
done in https://github.com/nerdbord/good-dev-hunting-app/pull/263/commits/3f28a890f5f76461c3c8546f281a796b84b202ec

Sumick reviewed (changes requested):
One blocking issue to fix - we are missing profileId update in session after profile creation

meczajkowski commented on src/auth.ts:
```diff
@@ -0,0 +1,157 @@
+      session.user.name = foundUser.profile?.fullName || ''
+      session.user.image = foundUser.avatarUrl ? foundUser.avatarUrl : ''
+      session.user.roles = foundUser.roles
+      session.user.profileId = foundUser.profile && foundUser.profile.id
```
done in https://github.com/nerdbord/good-dev-hunting-app/pull/263/commits/158d0e50a031291c831879b5aca6034c68077e5b

meczajkowski commented on src/app/(profile)/my-profile/create/page.tsx:
```diff
@@ -1,21 +1,19 @@
+  const { user, userIsHunter, userHasProfile } = await getAuthorizedUser()
 
-  if (!session?.user) {
+  console.log(user)
```
done in https://github.com/nerdbord/good-dev-hunting-app/pull/263/commits/158d0e50a031291c831879b5aca6034c68077e5b

Sumick reviewed (approved):
🔥🔥🔥🔥🔥

meczajkowski commented:
- Enhanced session management during profile creation and editing.
- Ensured user authentication with sessions in _actions.
- Simplified the User model to match the production model.
- Cleaned up migrations and removed the sessions model as we use JWT session strategy.
- Linked terms and privacy policy in the hunter login form.
- Eliminated the requireUserRoles, createGithubDetails method.



Authentication in components and pages now practically relies entirely on sessions. Thanks to this solution, I was able to significantly reduce database queries, resulting in much better performance.

Ready to merge!!!

mmcode9407 reviewed (approved):
Love ❤️
